tests/pattern-tests.js

```diff
@@ -504,4 +512,118 @@ function testPatterns(Prism) {
 		});
 	});
 
+	it('- should not cause polynomial backtracking', function () {
+		forEachPattern(({ pattern, ast, tokenPath }) => {
+			const patternStr = String(pattern);
+			if (polySafeRegexes.has(patternStr)) {
+				// we know that the pattern won't cause poly backtracking because we checked before
+				return;
+			}
+
+			const result = scslre.analyse(ast, { maxReports: 1, reportTypes: { 'Move': false } });
+			if (result.reports.length > 0) {
+				const polyCase = result.reports[0];
+
+				const lang = `/${polyCase.character.literal.source}/${polyCase.character.literal.flags}`;
+
+				let rangeOffset;
+				let rangeStr;
+				let rangeHighlight;
+				let exponential = polyCase.exponential;
+
+				switch (polyCase.type) {
+					case 'Trade': {
+						const start = Math.min(polyCase.startQuant.start, polyCase.endQuant.start);
+						const end = Math.max(polyCase.startQuant.end, polyCase.endQuant.end);
+						rangeOffset = start + 1;
+						rangeStr = patternStr.substring(start + 1, end + 1);
+						rangeHighlight = highlight([
+							{ ...polyCase.startQuant, label: 'start' },
+							{ ...polyCase.endQuant, label: 'end' }
+						], -start);
+						break;
+					}
+					case 'Self': {
+						rangeOffset = polyCase.parentQuant.start + 1;
+						rangeStr = patternStr.substring(polyCase.parentQuant.start + 1, polyCase.parentQuant.end + 1);
+						rangeHighlight = highlight([{...polyCase.quant, label: 'self'}], -polyCase.parentQuant.start);
+						break;
+					}
+					case 'Move': {
+						rangeOffset = 1;
+						rangeStr = patternStr.substring(1, polyCase.quant.end + 1);
+						rangeHighlight = highlight([polyCase.quant]);
+						break;
+					}
+					default:
+						throw Error();
```

Comment:
And maybe include a message? `"Should never happen :p"`

```suggestion
						throw new Error();
```